cranelift/peepmatic/crates/runtime/src/paths.rs

```diff
@@ -0,0 +1,242 @@
+//! Representing paths through the dataflow graph.
+//!
+//! Paths are relative from a *root* instruction, which is the instruction we
+//! are determining which, if any, optimizations apply.
+//!
+//! Paths are series of indices through each instruction's children as we
+//! traverse down the graph from the root. Children are immediates followed by
+//! arguments: `[imm0, imm1, ..., immN, arg0, arg1, ..., argN]`.
+//!
+//! ## Examples
+//!
+//! * `[0]` is the path to the root.
+//! * `[0, 0]` is the path to the root's first child.
+//! * `[0, 1]` is the path to the root's second child.
+//! * `[0, 1, 0]` is the path to the root's second child's first child.
+//!
+//! ## Interning
+//!
+//! To avoid extra allocations, de-duplicate paths, and reference them via a
+//! fixed-length value, we intern paths inside a `PathInterner` and then
+//! reference them via `PathId`.
+
+// TODO: Make `[]` the path to the root, and get rid of this redundant leading
+// zero that is currently in every single path.
+
+use serde::de::{Deserializer, SeqAccess, Visitor};
+use serde::ser::{SerializeSeq, Serializer};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+
+/// A path through the data-flow graph from the root instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Path<'a>(pub &'a [u8]);
+
+impl Path<'_> {
+    /// Construct a new path through the data-flow graph from the root
+    /// instruction.
+    pub fn new(path: &impl AsRef<[u8]>) -> Path {
+        Path(path.as_ref())
+    }
+}
+
+/// An identifier for an interned path.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct PathId(u32);
+
+/// An interner and de-duplicator for `Path`s.
+///
+/// Can be serialized and deserialized while maintaining the same id to interned
+/// path mapping.
+#[derive(Debug, Default)]
+pub struct PathInterner {
+    // A map from a path (whose owned data is inside `arena`) to the canonical
+    // `PathId` we assigned it when interning it.
+    map: HashMap<UnsafePath, PathId>,
+
+    // A map from a `PathId` index to an unsafe, self-borrowed path pointing
+    // into `arena`. It is safe to given these out as safe `Path`s, as long as
+    // the lifetime is not longer than this `PathInterner`'s lifetime.
+    paths: Vec<UnsafePath>,
+
+    // Bump allocation arena for path data. The bump arena ensures that these
+    // allocations never move, and are therefore safe for self-references.
```

Comment:
```suggestion
    /// A map from a path (whose owned data is inside `arena`) to the canonical
    /// `PathId` we assigned it when interning it.
    map: HashMap<UnsafePath, PathId>,

    /// A map from a `PathId` index to an unsafe, self-borrowed path pointing
    /// into `arena`. It is safe to given these out as safe `Path`s, as long as
    /// the lifetime is not longer than this `PathInterner`'s lifetime.
    paths: Vec<UnsafePath>,

    /// Bump allocation arena for path data. The bump arena ensures that these
    /// allocations never move, and are therefore safe for self-references.
```

These show up when using `--document-private-items`.